docs/proposals/20230612-node-os-image-caching.md

```diff
@@ -0,0 +1,230 @@
+
+##### FR1
+
+##### FR2
+
+#### Non-Functional Requirements
+
+##### NFR1
+
+##### NFR2
+
+### Implementation Details/Notes/Constraints
+
+The plan is to modify the existing Controllers with the Node Prototype Pattern as desired. These controller additions can be added to AzureMachine Controller and AzureMachinePool Controller.
+
+An operator will be able to decide to turn the feature on or off with an environment variable on creation of a cluster, and then can update the AzureMachineTemplate and AzureMachinePool to customize how long they want the caching interval to be (see the yaml files below in this section for the caching interval).
+
+Example of the enviornment variable being turned on:
+
+```
+export AZURE_OS_CACHING=true
+```
+
+The controller will maintain a timestamp, and when the current time is the chosen interval ahead or more, the controller will perform the caching. When the process is started it should go through the nodes of the cluster, choose a healthy node, shut it down, take a snapshot of it, restart it, create a shared image gallery image, delete the snapshot, and then configure the AzureMachineTemplate specs to use that shared image gallery image. After, it will store the current time as its timestamp.
+
+As for why the healthy node has to be shut down while creating a snapshot of it, if it isn’t shut down first then pods can be scheduled as the snapshot is taken which will cause some dangerous states in terms of how it exists after being utilized by the AzureMachineTemplates.
+
+In terms of how a healthy node would be selected, there is already an annotation on each node from its status as to when they’ve last been patched. An ideal node would be one which has been patched since the last prototype went into service and is running and healthy. Whichever node has been steadily healthy for the longest amount of time should be chosen as it’s the most overall stable. As the prototype is always from a successfully healthy and working node the image is always known to be working before being chosen for replication.
+
+In terms of when to take a snapshot, a day is given as a general example which should be good for typical use but the specification of how often will be customizable as we know that certain operators have different strategies and use cases for how they’re running their services on our clusters.
+
+In terms of data model changes, AzureMachine and AzureMachinePool changes are likely and the changes we expect will be purely additive and nonbreaking. No removals should be required to the data model. For AzureMachine on AzureMachineTemplate we will add a new optional field under spec.template.spec called nodeCachingMode which will be enabled if present and a sub field under that called cacheInterval which will be set to 1 day by default, and it will be a required subfield of nodeCachingMode if that field is there. For AzureMachinePool we will do the same thing under spec.template in the AzureMachinePool files.
```

Comment:
One thing this proposal hasn't explored yet is how to make the changes in a non-disruptive way for existing machines. The current default behavior for MachineDeployments and MachinePools is to trigger a rolling update if the template changes. This means all Machines will be automatically replaced so they always match the desired state. We don't want this every time the image is updated because it would be quite disruptive (imagine all your nodes having to be shutdown and rebuilt every day or however frequent the interval is). How can we avoid this?

Reply:
+1, I wonder if something like a new or existing `spec.strategy` from MachineDeployment might be valuable here: https://doc.crds.dev/github.com/kubernetes-sigs/cluster-api/cluster.x-k8s.io/MachineDeployment/v1beta1@v1.3.8#spec-strategy

Reply:
Mentioned how we can add an optional field to rollingUpdate and set it to false by default to prevent this automatic rollout